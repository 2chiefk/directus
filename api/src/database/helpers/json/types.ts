import { DatabaseHelper } from '../types';
import { Knex } from 'knex';
import { Item, SchemaOverview } from '@directus/shared/types';
import { JsonFieldNode } from '../../../types';

export abstract class JsonHelper extends DatabaseHelper {
	constructor(protected knex: Knex, protected schema: SchemaOverview, protected nodes: JsonFieldNode[] = []) {
		super(knex);
		this.schema = schema;
		this.nodes = nodes;
	}
	static isSupported(_version: string, _full = ''): boolean {
		return false;
	}

	abstract preProcess(dbQuery: Knex.QueryBuilder, table: string): void;
	abstract postProcess(items: Item[]): void;
}
