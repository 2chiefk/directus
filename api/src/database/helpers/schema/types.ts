import { getDatabaseClient } from '../../index';
import { KNEX_TYPES } from '@directus/shared/constants';
import { DatabaseClients, DatabaseHelper } from '../types';

export type Options = { nullable?: boolean; default?: any; length?: number };

export abstract class SchemaHelper extends DatabaseHelper {
	isOneOfClients(clients: DatabaseClients[]): boolean {
		return clients.includes(getDatabaseClient(this.knex));
	}

	async changeNullable(table: string, column: string, nullable: boolean): Promise<void> {
		await this.knex.schema.alterTable(table, (builder) => {
			if (nullable) {
				builder.setNullable(column);
			} else {
				builder.dropNullable(column);
			}
		});
	}

	async changeToType(
		table: string,
		column: string,
		type: typeof KNEX_TYPES[number],
		options: Options = {}
	): Promise<void> {
		await this.knex.schema.alterTable(table, (builder) => {
			const b = type === 'string' ? builder.string(column, options.length) : builder[type](column);

			if (options.nullable === true) {
				b.nullable();
			}

			if (options.nullable === false) {
				b.notNullable();
			}

			if (options.default !== undefined) {
				b.defaultTo(options.default);
			}

			b.alter();
		});
	}

	protected async changeToTypeByCopy(
		table: string,
		column: string,
		type: typeof KNEX_TYPES[number],
		options: Options
	): Promise<void> {
		const tempName = `${column}__temp`;

		await this.knex.schema.alterTable(table, (builder) => {
			const col = type === 'string' ? builder.string(tempName, options.length) : builder[type](tempName);

			if (options.default !== undefined) {
				col.defaultTo(options.default);
			}

			// Force new temporary column to be nullable (required, as there will already be rows in
			// the table)
			col.nullable();
		});

		await this.knex(table).update(tempName, this.knex.ref(column));

		await this.knex.schema.alterTable(table, (builder) => {
			builder.dropColumn(column);
		});

		await this.knex.schema.alterTable(table, (builder) => {
			builder.renameColumn(tempName, column);
		});

		// We're altering the temporary column here. That starts nullable, so we only want to set it
		// to NOT NULL when applicable
		if (options.nullable === false) {
			await this.changeNullable(table, column, options.nullable);
		}
	}

	async preColumnChange(): Promise<boolean> {
		return false;
	}

	async postColumnChange(): Promise<void> {
		return;
	}

	async getVersion(): Promise<{ parsed: number[]; full: string }> {
		const version = await this.knex.select(this.knex.raw('@@version'));
		return {
			parsed: version[0]['@@version'].split('.').map((num: string) => parseInt(num, 10)),
			full: version[0]['@@version'],
		};
	}
}
