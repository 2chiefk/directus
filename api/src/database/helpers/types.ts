import { Knex } from 'knex';

export type DatabaseClients = 'mysql' | 'postgres' | 'cockroachdb' | 'sqlite' | 'oracle' | 'mssql' | 'redshift';

export abstract class DatabaseHelper {
	constructor(protected knex: Knex) {}
}
