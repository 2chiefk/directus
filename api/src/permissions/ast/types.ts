export type CollectionKey = string;
export type FieldKey = string;
export type QueryPath = string[];

/**
 * Key is dot-notation QueryPath, f.e. `category.created_by`.
 * Value contains collection context for that path, and fields fetched within
 */
export type FieldMap = Map<string, { collection: CollectionKey; fields: Set<FieldKey> }>;

export interface AccessRow {
	policy: { id: string; admin_access: boolean; ip_access: string | null };
	sort: number;
	role: null | string;
	user: null | string;
}
