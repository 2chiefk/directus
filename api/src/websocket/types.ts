import type { Accountability, Query } from '@directus/shared/types';
import type { WebSocket } from 'ws';
import type { IncomingMessage } from 'http';
import type internal from 'stream';

export type AuthenticationState = {
	accountability: Accountability | null;
	expires_at: number | null;
	refresh_token?: string;
};

export type WebSocketClient = WebSocket & AuthenticationState & { uid: string; auth_timer: NodeJS.Timer | null };
export type UpgradeRequest = IncomingMessage & AuthenticationState;

export type Subscription = {
	uid?: string;
	query?: Query;
	item?: string | number;
	status?: boolean;
	collection: string;
	client: WebSocketClient;
};

export type UpgradeContext = {
	request: IncomingMessage;
	socket: internal.Duplex;
	head: Buffer;
};

export type ConnectionParams = {
	access_token?: string;
};

export type GraphQLSocket = {
	client: WebSocketClient;
};

export type AuthMode = 'public' | 'handshake' | 'strict';
