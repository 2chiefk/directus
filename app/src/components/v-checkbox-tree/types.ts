export type Choice = {
	text: string;
	value: string;
	children?: Choice[];
};
