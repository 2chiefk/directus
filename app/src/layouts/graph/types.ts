import { Filter } from '@directus/shared/types';

export type LayoutOptions = {
	relationField: string | null;
	fixedPositions: boolean;
	baseColor: string;
	baseSize: number;
	simulation: 'layout-force' | 'layout-forceatlas2'
	collectionsOptions: Record<string, CollectionOptions>;
};

export interface CollectionOptions {
	displayTemplate: string | null;
	colorField: string | undefined;
	sizeField: string | undefined;
	xField: string | undefined;
	yField: string | undefined;
	filters: FilterOptions[];
}

export type FilterOptions = {
	filter: Filter;
	color: string | undefined;
	size: number | undefined;
};

export type LayoutQuery = {
	fields: string[];
	sort: string[];
	page: number;
	limit: number;
};
