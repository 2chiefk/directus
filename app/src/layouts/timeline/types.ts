export type LayoutOptions = {
	dateField: string,
	timeField?: string,
	title: string,
	userField: string
};

export type Day = {
	date: string,
	events: Event[]
}

export type Event = {
	id: string | number,
	title: string,
	time?: string,
	user?: {
		id: string,
		first_name: string,
		last_name: string,
		image?: string,
		avatar?: {
			id: string,
			storage: string,
			filename_disk: string,
			type: string,
			modified_on: string
		}
	}
}

export type LayoutQuery = {
	fields?: string[];
	sort?: string;
	limit?: number;
	page?: number;
};
