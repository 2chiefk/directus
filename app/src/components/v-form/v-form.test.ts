import { test, expect, beforeEach, vitest, vi } from 'vitest';
import { GlobalMountOptions } from '@vue/test-utils/dist/types';
import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { createI18n } from 'vue-i18n';
import { Directive, h } from 'vue';

import VForm from './v-form.vue';

let global: GlobalMountOptions;

beforeEach(async () => {
	const i18n = createI18n();

	global = {
		plugins: [i18n, createTestingPinia()],
	};
});

vi.stubGlobal(
	'ResizeObserver',
	vi.fn(() => ({
		disconnect: vi.fn(),
		observe: vi.fn(),
		takeRecords: vi.fn(),
		unobserve: vi.fn(),
	}))
);

test('Mount component', () => {
	expect(VForm).toBeTruthy();

	const wrapper = mount(VForm, {
		global,
		props: {
			fields: [],
			collection: null,
			loading: false,
		},
		shallow: true,
	});

	expect(wrapper.html()).toMatchSnapshot();
});

test('Render text field', () => {
	// this correctly fails for the issue introduced by https://github.com/directus/directus/pull/15144
	const wrapper = mount(VForm, {
		global,
		props: {
			fields: [
				{
					collection: 'test',
					field: 'test',
					type: 'string',
					meta: {
						collection: 'test',
						field: 'test',
						special: null,
						interface: 'input',
						options: null,
						display: null,
						display_options: null,
						readonly: false,
						hidden: false,
						sort: null,
						width: 'full',
						translations: null,
						note: null,
						conditions: null,
						required: false,
						group: null,
						validation: null,
						validation_message: null,
					},
				},
			],
			collection: null,
			loading: false,
		},
		shallow: true,
	});

	expect(wrapper.html()).toMatchSnapshot();
});
