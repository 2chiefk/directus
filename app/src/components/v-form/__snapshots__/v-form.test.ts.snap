// Vitest Snapshot v1

exports[`Mount component 1`] = `
"<div class=\\"v-form grid\\" data-v-3fb8a5a2=\\"\\">
  <!--v-if-->
</div>"
`;

exports[`Render text field 1`] = `
"<div class=\\"v-form grid\\" data-v-3fb8a5a2=\\"\\">
  <!--v-if-->
  <form-field-stub field=\\"[object Object]\\" batchmode=\\"false\\" batchactive=\\"false\\" disabled=\\"false\\" loading=\\"false\\" autofocus=\\"false\\" raweditorenabled=\\"false\\" raweditoractive=\\"false\\" class=\\"first-visible-field\\" data-v-3fb8a5a2=\\"\\"></form-field-stub>
</div>"
`;
